### PR TITLE
[Wasm GC] Support and fuzz function subtyping

### PR DESCRIPTION
**Pull request by tlively**
Support function subtyping with contravariant parameters and covariant results.
The actual change is a single line in wasm-type.cpp, so most of the patch is
updating the type fuzzer to generate interesting function subtypes. Since
function parameters are covariant, generating a function subtype requires
generating supertypes of its parameter types, which required new functionality
in the fuzzer. Also update the fuzzer to choose to reuse types at a finer grain,
so for example individual function parameters or results might be reused
unmodified while other parameters or results are still modified.

### CODE REVIEW

**Comment by tlively** [comment]
Current dependencies on/for this PR:
* main
  * **PR #5420** <a href="https://app.graphite.dev/github/pr/WebAssembly/binaryen/5420" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/WebAssembly/binaryen/5420?utm_source=stack-comment).

File: src/tools/fuzzing/heap-types.cpp [hunk 1]
```diff
@@ -352,10 +410,76 @@ struct HeapTypeGeneratorImpl {
...
+    switch (type.getBasic()) {
+      case HeapType::ext:
+      case HeapType::func:
+      case HeapType::any:
+        break;
+      case HeapType::eq:
+        candidates.push_back(HeapType::any);
+        break;
+      case HeapType::i31:
+      case HeapType::struct_:
```
[the remark this answers is not in the document]
Comment (tlively) [reply src/tools/fuzzing/heap-types.cpp:460]: That's taken care of by `candidates.push_back(type)` on line 450 above.

File: test/lit/isorecursive-good.wast [hunk 2]
```diff
@@ -27,6 +34,11 @@
     (type $sub-array (array_subtype (ref $sub-struct) $super-array))
   )
 
+  (rec
+    (type $super-func (func (param (ref $sub-func)) (result (ref $super-func))))
+    (type $sub-func (func_subtype (param (ref $super-func)) (result (ref $sub-func)) $super-func))
+  )
```
[the remark this answers is not in the document]
Comment (tlively) [reply test/lit/isorecursive-good.wast:40]: Sure, will do.

File: src/tools/fuzzing/heap-types.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (kripken) [reply src/tools/fuzzing/heap-types.cpp:460]: Ah, I missed that, thanks.

File: test/lit/isorecursive-good.wast (hunk 2 above)
[the remark this answers is not in the document]
Comment (kripken) [reply test/lit/isorecursive-good.wast:40]: Cool, lgtm with that + fuzzing.

**Review by kripken — APPROVED** [review]

**Comment by tlively** [comment]
Looks like the fuzzer will require updating :'(

**Comment by tlively** [comment]
Fuzzed for a few tens of thousands of iterations and the only problem I've found is the unrelated bug with type merging that was fixed in #5424. I'll land this now and continue fuzzing just in case.

**Pull request merged**